pkg/scheduler/plugins/coescheduling/core/gang_cache.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	listerv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/scheduler-plugins/pkg/apis/scheduling/v1alpha1"
+
+	pglister "sigs.k8s.io/scheduler-plugins/pkg/generated/listers/scheduling/v1alpha1"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	"github.com/koordinator-sh/koordinator/apis/scheduling/config"
+)
+
+type GangCache struct {
+	lock       *sync.RWMutex
+	gangItems  map[string]*Gang
+	pluginArgs *config.CoschedulingArgs
+	podLister  listerv1.PodLister
+	pgLister   pglister.PodGroupLister
+}
+
+func NewGangCache(args *config.CoschedulingArgs, podLister listerv1.PodLister, pgLister pglister.PodGroupLister) *GangCache {
+	return &GangCache{
+		gangItems:  make(map[string]*Gang),
+		lock:       new(sync.RWMutex),
+		pluginArgs: args,
+		podLister:  podLister,
+		pgLister:   pgLister,
+	}
+}
+
+func (gangCache *GangCache) createGangToCache(gangId string) *Gang {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	if gang, exist := gangCache.gangItems[gangId]; exist {
+		return gang
+	}
+
+	gang := NewGang(gangId)
+	gangCache.gangItems[gangId] = gang
+	return gang
+}
+
+func (gangCache *GangCache) getGangFromCache(gangId string) *Gang {
+	gangCache.lock.RLock()
+	defer gangCache.lock.RUnlock()
+
+	return gangCache.gangItems[gangId]
+}
+
+func (gangCache *GangCache) deleteGangFromCache(gangId string) {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	delete(gangCache.gangItems, gangId)
+}
+func (gangCache *GangCache) getGangItems() map[string]*Gang {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	return gangCache.gangItems
+}
+
+func (gangCache *GangCache) RecoverGangCache() {
+	// recover podGroup First
+	pgLister := gangCache.pgLister
+	pgList, err := pgLister.List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("RecoverGangCache podGroupList List error: %v", err.Error())
+	}
+	for _, podGroup := range pgList {
+		gangCache.onPodGroupAdd(podGroup)
+
+	}
+	// recover gang from pod
+	podLister := gangCache.podLister
+	podsList, err := podLister.List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("RecoverGangCache podsList List error: %v", err.Error())
+	}
+	for _, pod := range podsList {
+		gangCache.onPodAdd(pod)
+
+		if pod.Spec.NodeName != "" {
```

Comment:
may be you can add a "recover bool" variable of onPodAdd, so here you can save get "gang" pointer's work.

Reply:
yes,done